doc/README.md

```diff
@@ -0,0 +1,421 @@
+      "length": 26549052,
+      "uncompressedLength": 28549052
+    },
+    "c11c3cf1c326375b177e768de908e6e423c3d8866715c8ec7159156b6ea82497": {
+      "id": "5421895dbead0ba83f0993f17b5f72f6c6618c0e9ea849c6c53ac240d1802d0b",
+      "length": 523496,
+      "uncompressedLength": 645314
+    }
+  }
+}
+```
+
+The `chunkIds` and `iconChunkIds` fields contain a list with plain text SHA-256 hashes
+which can be found in the main `blobs` dictionary.
+This contains a mapping from plain text SHA-256 hashes to storage IDs and size information.
+
+At the beginning of most operations, we download all available snapshots
+to get information about all blobs that should be available in the repository.
+We may additionally retrieve a list of all blobs directly from the repository
+to ensure they are actually (still) present.
+
+Snapshot files start with the SHA-256 hash of the content of the file
+and use a `.snapshot` extension.
+
+## Cryptography
+
+This section is based on and thus very similar to encryption of
+[files backup](../storage/doc/design.md).
+
+### Main Key
+
+Seedvault already uses [BIP39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki)
+to give users a mnemonic recovery code and for generating deterministic keys.
+The derived key has 512 bits
+and Seedvault used to use the first 256 bits as an AES key to encrypt app data.
+Unfortunately, this key's usage is limited by Android's keystore to encryption and decryption.
+Therefore, the second 256 bits are be imported into Android's keystore for use with `HMAC-SHA256`,
+so that this key can act as a main key we can deterministically derive additional keys from
+by using HKDF ([RFC5869](https://tools.ietf.org/html/rfc5869)).
+These second 256 bits must not be used for any other purpose in the future.
+We use them for a main key to avoid users having to handle yet another secret.
+
+For deriving keys, we are only using the HKDF's second 'expand' step,
+because the Android Keystore does not give us access
+to the key's byte representation (required for first 'extract' step) after importing it.
+This should be fine as the input key material is already a cryptographically strong key
+(see section 3.3 of RFC 5869 above).
+
+### Key derivation overview
+
+The original entropy comes from a BIP39 seed (12 words = 128 bit size)
+obtained from Java's `SecureRandom`.
+A PBKDF SHA512 based derivation defined in BIP39 turns this into a 512 bit seed key.
+
+The derived seed key (512 bit size) gets split into two parts:
+1. legacy app data encryption key (unused) - 256 bit - first half of seed key
+2. main key - 256 bit - second half of seed key used to derive application specific keys:
+   1. App Backup: HKDF with info "app backup repoId key"
+   2. App Backup: HKDF with info "app backup gear table key"
+   3. App Backup: HKDF with info "app backup stream key"
+   4. File Backup: HKDF with info "stream key"
+   5. File Backup: HKDF with info "Chunk ID calculation"
+
+### Stream Encryption
```

Comment:
[Clarity]
I suspect that a "stream" refers to an input or output stream that would typically always be backed by a file...? Is there an important distinction between "streams" and "files"? Should this document *always* be talking about streams instead of files, or vice versa?

The Data Format section refers to "files". This is our first introduction to "streams" (it's not in the definitions).

If the distinction is important or otherwise intentional, then as with the above comment, there appears to be some duplicate information on the lines below - in this case from the Data Format section - so consider acknowledging that it was referenced earlier. Or perhaps the sections should be merged in some way...?

Reply:
Okay, based on that explanation, I think this is fine as-is.